presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -314,18 +314,20 @@ public void flushCache()
     private List<String> loadAllDatabases()
             throws Exception
     {
+        Optional<IMetaStoreClient> metastoreClient = Optional.of(clientProvider.createMetastoreClient());
         try {
             return retry()
                     .stopOnIllegalExceptions()
                     .run("getAllDatabases", stats.getGetAllDatabases().wrap(() -> {
-                        try (HiveMetastoreClient client = clientProvider.createMetastoreClient()) {
```

Comment:
The caching and retry logic is generally beneficial which is why I took this approach. I also took this approach because it allows someone to not only use a different metastore implementation easily with Presto (by just updating the factory) but also with Hive.